Wrappers/Python/cil/optimisation/algorithms/SIRT.py

```diff
@@ -97,22 +134,51 @@ def set_up(self, initial, operator, data, lower=None, upper=None, constraint=Non
                 self.constraint=IndicatorBox(lower=lower,upper=upper)
                 
         # Set up scaling matrices D and M.
-        self.M = 1/self.operator.direct(self.operator.domain_geometry().allocate(value=1.0))        
-        self.D = 1/self.operator.adjoint(self.operator.range_geometry().allocate(value=1.0))
+        self.M = 1./self.operator.direct(self.operator.domain_geometry().allocate(value=1.0))                
+        self.D = 1./self.operator.adjoint(self.operator.range_geometry().allocate(value=1.0))
+
+        # fix for possible inf values (code from numpy.nan_to_nam)
+        # TODO replace with
+        # numpy.nan_to_num(self.M, copy = False, nan = 1, neginf=1, posinf=1) 
+        # numpy.nan_to_num(self.D, copy = False, nan = 1, neginf=1, posinf=1) 
+
+        idx_nan1 = numpy.isnan(self.M.as_array())
+        idx_pinf1 = numpy.isposinf(self.M.as_array())
+        idx_ninf1 = numpy.isneginf(self.M.as_array())        
+        numpy.copyto(self.M.as_array(), 1., where=idx_nan1)
+        numpy.copyto(self.M.as_array(), 1., where=idx_pinf1)
+        numpy.copyto(self.M.as_array(), 1., where=idx_ninf1)
+
```

Comment:
`numpy.isnan() `,  `numpy.isposinf()` and  `numpy.isneginf()` all return boolean arrays, not lists of indices. so better to 'or' them first and just loop through the array replacing values once, as you replace everything with a 1.

I'm then confused by `copyto`
https://numpy.org/doc/1.15/reference/generated/numpy.copyto.html?highlight=copyto

Your call doesn't seem to match the signature, I'd expect it to copy from src to dst only where `where=True`. In your call you don't have a `dst`, instead you pass 1, I struggle to see how this is doing what you want.

I would imaging what you want is just:
```
arr = M.as_array()
arr_replace = numpy.isnan(arr) or  numpy.isposinf(arr) or numpy.isneginf(arr)
arr [arr_replace] = 1
M.fill(arr)
```

Reply:
But also, if it's a few indices that need replacing a boolean array is a pretty inefficient way to replace them, so this should really be the backup behaviour where `num_to_nan` can't be used.

Reply:
I use the code from [numpy.1.15](https://github.com/numpy/numpy/blob/ccfbcc1cd9a4035a467f2e982a565ab27de25b6b/numpy/lib/type_check.py#L404-L409)
The same code, with the `nan`, `posinf` and `neginf` is used in [numpy.1.22](https://github.com/numpy/numpy/blob/4adc87dff15a247e417d50f10cc4def8e1c17a03/numpy/lib/type_check.py#L514-L521)

The value 1, is broadcasted.

Reply:
I don't think we should rely on undocumented behaviour.  `copyto` may have changed between versions, and with the documentation not showing the behaviour you use I think that is very dangerous. particularly when another simple solution exists.

Secondly on this with cased where as_array() returns an array rather than a reference (i.e. sirf objects) this will fail 
`numpy.copyto(self.M.as_array(), 1., where=idx_nan1)`

As the modified M is discarded. You must use `as_array` paired with `fill` if you are going to rely on numpy operations.

Reply:
But [`fill`](https://github.com/TomographicImaging/CIL/blob/master/Wrappers/Python/cil/framework/framework.py#L2191) is a `numpy.copyto` wrapper

Reply:
I'm refereeing to `fill` from SIRF. The code will work with CIL objects. the `as_array`/`fill` is to support SIRF.